camel/messages/axolotl/sharegpt/functions/hermes/hermes_function_formatter.py

```diff
@@ -0,0 +1,79 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import json
+import re
+from typing import Any, Dict, List, Optional
+
+from camel.messages.axolotl.sharegpt.functions.function_call_formatter import (
+    FunctionCallFormatter,
+)
+from camel.messages.axolotl.sharegpt.functions.tool_call import ToolCall
+from camel.messages.axolotl.sharegpt.functions.tool_response import (
+    ToolResponse,
+)
+
+
+class HermesToolResponse(ToolResponse):
+    """Represents a single tool/function call with validation"""
+
+
+class HermesToolCall(ToolCall):
+    """Represents a single tool/function call with validation"""
+
+
+class HermesFunctionFormatter(
+    FunctionCallFormatter[HermesToolCall, HermesToolResponse]
+):
+    """Hermes-style function calling format implementation with validation"""
+
+    def extract_tool_calls(self, message: str) -> List[HermesToolCall]:
+        tool_calls = []
+        pattern = r"<tool_call>\s*({.*?})\s*</tool_call>"
+        matches = re.finditer(pattern, message, re.DOTALL)
+
+        for match in matches:
+            try:
+                call_dict = json.loads(match.group(1).replace("'", '"'))
+                tool_calls.append(HermesToolCall.model_validate(call_dict))
+            except Exception as e:
+                print(f"Warning: Failed to parse tool call: {e}")
+                continue
+
+        return tool_calls
+
+    def extract_tool_response(
```

Comment:
Yeah these 2 methods are for when you want to get tool information from strings, likely but not necessarily part of message content. I think its good to have them in this class whose responsibility is about reading and writing this function calling format (to be clear, the <tool_response> part isn't a regex group but looking for that string literally because it marks where a tool's response json is in the hermes function calling format. A more robust way could be to parse the string as xml and navigate the xml element tree, but that just felt like overkill for now to do in a way that wouldn't break from invalid xml)